drivers/adc/adc_stm32.c

```diff
@@ -984,12 +1026,25 @@ static void adc_stm32_isr(const struct device *dev)
 	ADC_TypeDef *adc = config->base;
 
 	*data->buffer++ = LL_ADC_REG_ReadConversionData32(adc);
-
-	adc_context_on_sampling_done(&data->ctx, dev);
+	data->samples_count++;
+	/* ISR is triggered after each conversion, and at the end-of-sequence. */
+	if (data->samples_count % data->channel_count == 0) {
```

Comment:
Since `samples_count` is just 8 bits, the counter will possibly overflow.  This doesn't matter if `channel_count` is a power of two, but it doesn't need to be, and if it's not a power of two the result won't be correct.

I suggest:
```
if (data->samples_count == data->channel_count) {
        data->samples_count = 0;
}
```
Firstly, it easily avoids the overflow problem.  Secondly, the modulus (`data->channel_count`) is non-constant.  The compiler can come up with efficient code to find a constant modulus.  Obviously for a power of two it's just an AND operation, but even for numbers like 5 there is fast code.  But for a non-constant it has to do full integer division, which is very slow.  Usually, there is no integer division instruction (ARM), but even when there is (x86), it's by far the slowest ALU instruction in the CPU.  Even slower than floating point division surprisingly.

Reply:
Changed. Good catch :)